python/semantic_kernel/connectors/memory/mongodb_atlas/mongodb_atlas_memory_store.py

```diff
@@ -0,0 +1,342 @@
+        self.__index_name = index_name or DEFAULT_SEARCH_INDEX_NAME
+        self.__num_candidates = num_candidates or DEFAULT_NUM_CANDIDATES
+
+    @property
+    def database_name(self) -> str:
+        return self.__database_name
+
+    @property
+    def database(self) -> core.AgnosticDatabase:
+        return self._mongo_client[self.database_name]
+
+    @property
+    def index_name(self) -> str:
+        return self.__index_name
+
+    @property
+    def num_candidates(self) -> int:
+        return self.__num_candidates
+
+    async def close_async(self):
+        """Async close connection, invoked by MemoryStoreBase.__aexit__()"""
+        if self._mongo_client:
+            self._mongo_client.close()
+            self._mongo_client = None
+
+    async def create_collection_async(self, collection_name: str) -> None:
+        """Creates a new collection in the data store.
+
+        Arguments:
+            collection_name {str} -- The name associated with a collection of embeddings.
+
+        Returns:
+            None
+        """
+        if not await self.does_collection_exist_async(collection_name):
+            await self.database.create_collection(collection_name)
+
+    async def get_collections_async(
+        self,
+    ) -> List[str]:
+        """Gets all collection names in the data store.
+
+        Returns:
+            List[str] -- A group of collection names.
+        """
+        return await self.database.list_collection_names()
+
+    async def delete_collection_async(self, collection_name: str) -> None:
+        """Deletes a collection from the data store.
+
+        Arguments:
+            collection_name {str} -- The name associated with a collection of embeddings.
+
+        Returns:
+            None
+        """
+        await self.database[collection_name].drop()
+
+    async def does_collection_exist_async(self, collection_name: str) -> bool:
+        """Determines if a collection exists in the data store.
+
+        Arguments:
+            collection_name {str} -- The name associated with a collection of embeddings.
+
+        Returns:
+            bool -- True if given collection exists, False if not.
+        """
+        return collection_name in (await self.get_collections_async())
+
+    async def upsert_async(self, collection_name: str, record: MemoryRecord) -> str:
+        """Upserts a memory record into the data store. Does not guarantee that the collection exists.
+            If the record already exists, it will be updated.
+            If the record does not exist, it will be created.
+
+        Arguments:
+            collection_name {str} -- The name associated with a collection of embeddings.
+            record {MemoryRecord} -- The memory record to upsert.
+
+        Returns:
+            str -- The unique identifier for the memory record.
+        """
+        await self.create_collection_async(collection_name)
+
+        if not record._id:
+            record._id = str(uuid.uuid4())
```

Comment:
@Jibola does the ID of a mongoDB vector record have to a UUID? 

Ideally, the unique ID for a vector used by SK is a natural language string (see python/samples/kernel-syntax-examples/memory.py). We realize that some vector stores don't support this though, and we've had to implement some interesting logic to get around this - see qdrant_memory_store.py. Any memory connector should be able to seemlessly replace VolatileMemoryStore - with the exception of index creation which may have to be done outside of SK.

Reply:
No, it doesn't need to be a uuid at all. I had been working off assumptions seen from other implementations. My question then is, _is record._id guaranteed?_. 
* If yes, then I can remove this section outright

* If not, then my next question would be _does it have to be unique?_
  * If yes, then we'd have to make _something_ the string representation or throw an exception so the end-user knows
  * If not, then we can demote id from being the primary key in the document, and populate and arbitrary--somewhat untracked--ObjectId to be the "primary key" and would be obfuscated from code. 

Let me know what version works best here

Reply:
```
class MemoryRecord:
    _key: str
    _timestamp: Optional[datetime]
    _is_reference: bool
    _external_source_name: Optional[str]
    _id: str
    _description: Optional[str]
    _text: Optional[str]
    _additional_metadata: Optional[str]
    _embedding: ndarray
```

Pasting the attributes of MemoryRecord above for reference. 

Here are a list of scenarios and the expectations for each: 

1. The data is completely new (insert):
**In this scenario, MemoryRecord._key will be empty. When creating the mongoDB record, MemoryRecord._key = MemoryRecord._id. For vector stores that do not support natural language keys, MemoryRecord._key != MemoryRecord._id, but it sounds like we don't have to worry about that for Atlas :)**

2. The data is not new (intentional update):
**In this scenario, MemoryRecord._key will be populated. Upsert will act as an update.**

3. The data is intended to be new, but there's a key collision (unintentional update):
**This scenario is identical to scenario 2, but unintentional. We recognize that key collisions are more likely to happen when using natural language as keys. We intentionally called this part of SK 'semantic memory' to associate what it stores with relevance as opposed to a dumping ground of information. Ensuring that keys can be searchable and 'relevant' is an important aspect of our vision for SK. It does come with different assumptions than traditional big data storage, and it It is up to the app layer to work with those assumptions.**

Reply:
cool thank you!